optlib/terraform.ctags

```diff
@@ -0,0 +1,42 @@
+#
+# terraform.ctags: regex parse for Terraform (HCL).
+#
+# Copyright (c) 2021, Julio Tain Sueiras
+# Copyright (c) 2021, Antony Southworth
+#
+# Antony Southworth adapted Julio's code to Universal Ctags.
+#
+# This source code is released for free distribution under the terms of the
+# GNU General Public License version 2 or (at your option) any later version.
+#
+# Derived from `vim-terraform-completion`:
+#   - https://github.com/juliosueiras/vim-terraform-completion/blob/master/ctags/terraform.ctags
+#
+# Changed the name from `terraform` to `tf` so vim will recognise it properly based
+# on file extension (*.tf).
+#
+# A notable abscence is `local`, because `locals` are defined inside a block and
+# it's way harder to write a good regex for that.
+#   - https://www.terraform.io/docs/language/values/locals.html
+
+--langdef=Terraform
+
+--map-Terraform=+.tf
+--map-Terraform=+.tfvars
+
+--kinddef-Terraform=r,resource,Terraform Resource
+--kinddef-Terraform=d,data,Terraform Data
+--kinddef-Terraform=v,variable,Terraform Variable
+--kinddef-Terraform=p,provider,Terraform Provider
+--kinddef-Terraform=m,module,Terraform Module
+--kinddef-Terraform=o,output,Terraform Output
+
+--regex-Terraform=/^resource[[:space:]]"([^"]+)"[[:space:]]"([^"]+)"/\2/r/
+--regex-Terraform=/^data[[:space:]]"([^"]+)"[[:space:]]"([^"]+)"/\2/d/
+--regex-Terraform=/^variable[[:space:]]"([^"]+)"/\1/v/
+--regex-Terraform=/^provider[[:space:]]"([^"]+)"/\1/p/
+--regex-Terraform=/^module[[:space:]]"([^"]+)"/\1/m/
+--regex-Terraform=/^output[[:space:]]"([^"]+)"/\1/o/
+
+# ???
+--regex-Terraform=/^([a-z0-9_]+)[[:space:]]*=/\1/f,TFVar/
```

Comment:
@Haggus What do you think about this?

It seems that this is not a variable definition.
Definitions of variables are extracted with `v`.

Is `TFVar` a different thing from `v`?

Reply:
So in Terraform, you can define variables (`variable` keyword, or `v` in this case), but then you can use "*.tfvars" file to overwrite them.

For example, let's say you have a virtual machine instance which is named "user-service":
```hcl
variable "user_service_name" {
  type = string
}

resource "aws_instance" "user-service" {
  name = var.user_service_name
}
```

Then for different environments, you can overwrite this variable. For example with "dev.tfvars" file:
```hcl
user_service_name = "dev-user-service"
```

In this case, when searching for tags in "dev.tfvars", `user_service_name` should point the `variable "user_service_name" {` line.

I hope this makes sense. I'm new to the internals of ctags, so I any changes I can make, just let me know. I'm open to suggestions.

Reply:
I should've added a test case for this

Reply:
Thank you. Interesting. As far as reading your comment .tf and .tfvars files are not the same things. I think we should introduce just another parser: TFVar parser, which defines only one kind.

```
--langdef=TFVar
--map-TFVar=.tfvars
--kinddef-TFVar=o,overwritevars,the names for overwriting variables
--regex-TFVar=/^([a-z0-9_]+)[[:space:]]*=/\1/o/
```

We we can remove `--regex-Terraform=/^([a-z0-9_]+)[[:space:]]*=/\1/f,TFVar/` from the Terraform parser.

Reply:
The ideal way is to introduce the TerraformVariables parser and let the parser emit Terraform's variable kind tags with "overwritten" role.
A bit complicated but clean. I will try to implement the TerraformVariables parser. Give me some time.